### ISSUE TITLE
Change in behavior for GenericAlias

### ISSUE THREAD

**Issue by adampauls** [microsoft/pyright#5767]
**Describe the bug**
As of 1.1.323:

```python
from types import GenericAlias
from typing import TypeVar


def substitute_type_variables(
    t: type | TypeVar | GenericAlias,
    substitutions: dict[TypeVar, type],
) -> type | TypeVar | GenericAlias:
    raise NotImplementedError


def test_substitute_type_variables():
    T = TypeVar("T")
    substitutions = {T: int}
    x: list[T] | None = None # error: Type variable "T" has no meaning in this context
    print(x)
    assert substitute_type_variables(list[T], substitutions) == list[int]  # error: Type variable "T" has no meaning in this context
```
The first error is correct and was there as of 1.1.322. The second error is new as of 1.1.323 and (I think) not desirable because code operating on types at runtime should be able to take GenericAliases as arguments. But maybe it's intended behavior? 

**VS Code extension or command-line**
Command line.

**Comment by erictraut** [comment]
This behavior is correct. I recently added a specific error to identify incorrect usage of type variables. Previously, this was going undetected. Your code sample is using `T` in a manner that is illegal and meaningless. Type variables in Python must be associated with ("bound to") the scope of a generic function, class, or type alias. In your case, `T` is not bound to any of these, so it has no valid meaning in this context.

**Comment by adampauls** [comment]
Well, this code executes and does something meaningful. It's just operating on types at the meta level. If I want to write a decorator like

```python

@change_return_type(int)
def my_return_type(x: T) -> list[T]:
    ...
```

I could do that by calling the code above on the result of `inspect.signature`. 

If I want to write and test code that operates at this meta-level, how should I do it? What is the right way to construct the GenericAlias for `list[T]` in a unit test?

**Comment by adampauls** [comment]
Sorry, I realize the above is a bad example. We have written a library that uses type annotations for serialization and we make heavy use of `inspect.signature` to map codecs of fields to their registered codecs. In order to handle generic type annotations, we need to be able to process 
```python
@dataclass
class Foo(Generic[T]):
   x: list[T]
```

We only need to construct GenericAliases in unit tests to test utility code, but it's still meaningful to do so. What's the preferred way? Or should we just resort to `pyright: ignore`?

**Comment by erictraut** [comment]
There are infinite examples of code that "runs and does something meaningful" but does not pass type checking, so that's not a useful criterion. ;)

If you want your tests to pass type checking, they'll need to be consistent with the Python static type system and its rules. For example, you could use a function-scoped type variable like this:

```python
T = TypeVar("T")

def test_substitute_type_variables(v: T) -> T:
    substitutions = {T: int}
    x: list[T] | None = None
    print(x)
    assert substitute_type_variables(list[T], substitutions) == list[int]
    return v
```

**Comment by adampauls** [comment]
I'll also add that `assert substitute_type_variables(T, substitutions) == int` in the above works just fine, so I think the check is inconsistent at this point. 

I think the better workaround is 
```python
assert substitute_type_variables(GenericAlias(list, T), substitutions) == list[int] 
```

Though even that might not work if you change the check to catch the bare type variable case as well. 

Thanks for the clarification!

**Comment by adampauls** [comment]
Oh, nevermind, my workaround doesn't work. `T` loses its identity that way (it becomes `~T`). 

I'm not sure how your workaround works in a unit test, since in order to instantiate the fixture `v`, you'd have to make a fixture that returns a `T`, but you can't do that without making some make value and casting it to a `T` I Ithink?

**Comment by adampauls** [comment]
`GenericAlias(list, T)` does work most of the time, though with some issues. I didn't realize that `GenericAlias(tuple, (T, U))` and `GenericAlias(tuple, [T, U]))` are not the same, which was causing my first issue above.

**Comment by adampauls** [comment]
I'll make one last case for undoing this change:

Suppose I have 
```python
def foo(t: TypeVar) -> None:
    ....
```

Should all such functions be banned? If the answer is "no", then be able to call `foo(TypeVar("T"))`? I think so, since how else could I call it? If so, should I then be able to call
```python
T = TypeVar("T")
foo(T)
```
I would think so?

My sense is that when you say that `T` is "meaningless" in that context, you are talking about `T` as a type variable in the static type checker. That's unquestionably true. But `T` as just any old Python value that happens to point to a `TypeVar` (which is just a regular class according to the Python runtime) seems like it should be fine? `list[T]` is just the way you write out `GenericAlias` literals in Python, it's not a static type.

**Comment by erictraut** [comment]
This isn't going to be changed. You're confusing static type constructs with runtime implementation details that happen to be used to implement these constructs.

If you're writing specialized code that needs to understand the under-the-hood runtime implementation details of the static type system (things like the runtime type of a `TypeVar` or `GenericAlias`), you will need to make accommodations if you want your code to pass static type checking.

The errors that pyright emits are in place to aid developers who are using TypeVars for their intended use as described in PEP 484.

**Comment by adampauls** [comment]
Sorry, I can't resist: I am definitely not confusing the two! The check is. `list[T]` in `substitute_type_variables(list[T], substitutions)` is _not_ PEP 484 type hint, it's just an indexing expression (well, I think, I'm not totally sure what's going on under the hood, but I think [this](https://github.com/python/typeshed/blob/main/stdlib/builtins.pyi#L980) is the relevant stub?). Python type hints are so confusing because the static and runtime worlds interact, though thankfully that will be (largely) fixed by PEP 695. I think that might actually make the situation more clear. Suppose I have this in Python 3.12:

```python
def foo[T](x: T, y: TypeVar) -> None:
    good1: list[T] = list(x)
    good2: GenericAlias = list[y]
    bad1: list[y] = list(x)
    bad2: GenericAlias = list[T]
```

As of 1.1.323, pyright will emit on error about `bad1` and `good2`. I'm not sure what it would do with `bad2`, but I think it might _not_ rule it out because `list[T]` is a fine type hint, but a meaningless runtime expression (just as `list[y]` is a meaningless type hint but a fine runtime expression).

I think you're saying that any use of TypeVars outside of type hints -- even 100% correct code that must manipulate `TypeVar`s at runtime that come from the output `inspect.signature` -- will have to make accommodations to pass static type checking. If it were up to me, `list[str]` would not be the way to write `GenericAlias(list, (str,))` in Python since I think that is quite confusing, but here we are!

**Comment by adampauls** [comment]
Er, I guess even better: if `TypeVar` were not a runtime construct at all, and `inspect.signature` returned a different runtime class like `RuntimeTypeVar`, then that would eliminate a lot of confusion. In that case, the new check would be 100% correct.
